crates/bevy_ecs/src/storage/resource_non_send.rs

```diff
@@ -0,0 +1,605 @@
+        let task: Box<dyn FnOnce(&mut ThreadLocals) + Send + 'static> =
+            // SAFETY: This function will block the calling thread until `f` completes,
+            // so any captured references in `f` will remain valid until then.
+            unsafe { std::mem::transmute(task) };
+
+        // Send task to the main thread.
+        sender
+            .send_task(task)
+            .unwrap_or_else(|_| panic!("receiver missing"));
+
+        // Wait to receive result back from the main thread.
+        match result_rx.recv().unwrap() {
+            Ok((result, last_run)) => {
+                *self.last_run = last_run;
+                result
+            }
+            Err(payload) => {
+                std::panic::resume_unwind(payload);
+            }
+        }
+    }
+}
+
+// SAFETY: This system param does not borrow any data from the world.
+unsafe impl SystemParam for ThreadLocal<'_, '_> {
+    type State = ThreadLocalState;
+    type Item<'w, 's> = ThreadLocal<'w, 's>;
+
+    fn init_state(
+        world: &mut crate::prelude::World,
+        system_meta: &mut crate::system::SystemMeta,
+    ) -> Self::State {
+        // `ThreadLocalTaskSender` can't require `Sync` because `winit`'s `EventLoopProxy` is not
+        // `Sync`, so we need exclusive access to `ThreadLocalChannel` here.
+        let component_id =
+            crate::system::ResMut::<ThreadLocalChannel>::init_state(world, system_meta);
+
+        ThreadLocalState {
+            component_id,
+            last_run: Tick::new(0),
+        }
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &crate::system::SystemMeta,
+        world: UnsafeWorldCell<'world>,
+        curr_tick: Tick,
+    ) -> Self::Item<'world, 'state> {
+        let mut accessor = crate::system::ResMut::<ThreadLocalChannel>::get_param(
+            &mut state.component_id,
+            system_meta,
+            world,
+            curr_tick,
+        );
+
+        let access = if std::thread::current().id() == accessor.owning_thread {
+            let arc = accessor
+                .direct
+                .upgrade()
+                .expect("pointer to `ThreadLocals` should be valid");
+
+            // Use a raw pointer so we can hold onto the `Arc` and satisfy the borrow checker.
+            let ptr = Arc::as_ptr(&arc);
+
+            // SAFETY: This pointer is valid since we're still holding the `Arc`.
+            let mutex = unsafe { &*ptr };
+            let guard = mutex
+                .try_lock()
+                .expect("lock on `ThreadLocals` should be available");
+
+            ThreadLocalAccess::Direct(ThreadLocalsGuard { guard, _arc: arc })
+        } else {
+            let ptr: *mut dyn ThreadLocalTaskSender = &mut *accessor.indirect;
+            // SAFETY: The pointer is valid. We have to do this to satisfy the borrow checker.
+            ThreadLocalAccess::Indirect(unsafe { &mut *ptr })
```

Comment:
Couldn't this be `ThreadLocalAccess::Indirect(&mut *accessor.into_inner().indirect)`?

Reply:
Thanks! Nice catch.